File: extension/shared/prompt-template.js
(function (global) {
    'use strict';

    const PLACEHOLDER_PATTERN = /\{\{([^}]*)\}\}/g;

    function analyzeVariables(text) {
        const variables = [];
        const emptyPlaceholders = [];
        const seen = new Set();
        let match;

        PLACEHOLDER_PATTERN.lastIndex = 0;
        while ((match = PLACEHOLDER_PATTERN.exec(String(text))) !== null) {
            const name = match[1].trim();

            if (!name) {
                emptyPlaceholders.push(match[0]);
                continue;
            }

            if (!seen.has(name)) {
                seen.add(name);
                variables.push(name);
            }
        }

        return { variables, emptyPlaceholders };
    }

    function extractVariables(text) {
        return analyzeVariables(text).variables;
    }

    function substituteVariables(text, values) {
        const source = String(text);
        const normalizedValues = values && typeof values === 'object' ? values : {};
        const unfilled = new Set();

        PLACEHOLDER_PATTERN.lastIndex = 0;
        const result = source.replace(PLACEHOLDER_PATTERN, (placeholder, rawName) => {
            const name = rawName.trim();

            if (!name) {
                return placeholder;
            }

            const value = normalizedValues[name];
            if (typeof value !== 'string' || !value.trim()) {
                unfilled.add(name);
                return placeholder;
            }

            return value.trim();
        });

        return { text: result, unfilled: Array.from(unfilled) };
    }

    global.PromptTemplate = Object.freeze({
        analyzeVariables,
        extractVariables,
        substituteVariables
    });
})(window);

File: launcher/PromptLauncher.Core/PromptTemplate.cs
using System.Text.RegularExpressions;

namespace PromptLauncher.Core;

public static partial class PromptTemplate
{
    [GeneratedRegex(@"\{\{([^}]*)\}\}")]
    private static partial Regex PlaceholderPattern();

    public sealed record Analysis(IReadOnlyList<string> Variables, IReadOnlyList<string> EmptyPlaceholders);

    public static Analysis Analyze(string? text)
    {
        var variables = new List<string>();
        var empty = new List<string>();
        var seen = new HashSet<string>(StringComparer.Ordinal);

        foreach (Match match in PlaceholderPattern().Matches(text ?? ""))
        {
            var name = match.Groups[1].Value.Trim();
            if (name.Length == 0) empty.Add(match.Value);
            else if (seen.Add(name)) variables.Add(name);
        }
        return new(variables, empty);
    }

    public static SubstitutionResult Substitute(string? text, IReadOnlyDictionary<string, string>? values)
    {
        var unfilled = new List<string>();
        var seen = new HashSet<string>(StringComparer.Ordinal);
        var result = PlaceholderPattern().Replace(text ?? "", match =>
        {
            var name = match.Groups[1].Value.Trim();
            if (name.Length == 0) return match.Value;
            if (values is null || !values.TryGetValue(name, out var value) || string.IsNullOrWhiteSpace(value))
            {
                if (seen.Add(name)) unfilled.Add(name);
                return match.Value;
            }
            return value.Trim();
        });
        return new(result, unfilled);
    }
}
